File: sns_rooster/lib/services/cache_service.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:flutter/foundation.dart';

/// Cache Service for Flutter App
/// Provides intelligent caching to reduce API calls and improve performance

class CacheItem<T> {
  final T data;
  final DateTime timestamp;
  final Duration ttl;
  final String key;

  CacheItem({
    required this.data,
    required this.timestamp,
    required this.ttl,
    required this.key,
  });

  bool get isExpired => DateTime.now().difference(timestamp) > ttl;

  Map<String, dynamic> toJson() {
    return {
      'data': data,
      'timestamp': timestamp.toIso8601String(),
      'ttl': ttl.inMilliseconds,
      'key': key,
    };
  }

  factory CacheItem.fromJson(
      Map<String, dynamic> json, T Function(dynamic) fromJson) {
    return CacheItem<T>(
      data: fromJson(json['data']),
      timestamp: DateTime.parse(json['timestamp']),
      ttl: Duration(milliseconds: json['ttl']),
      key: json['key'],
    );
  }
}

class CacheService {
  static const String _cachePrefix = 'app_cache_';
  static const String _statsKey = 'cache_stats';

  // Cache configurations
  static const Duration shortTTL = Duration(seconds: 30);
  static const Duration mediumTTL = Duration(minutes: 5);
  static const Duration longTTL = Duration(minutes: 30);
  static const Duration staticTTL = Duration(hours: 24);

  // Cache statistics
  int _hits = 0;
  int _misses = 0;
  int _sets = 0;

  /// Set an item in cache
  Future<void> set<T>(
    String key,
    T data,
    Duration ttl, {
    T Function(dynamic)? fromJson,
    dynamic Function(T)? toJson,
  }) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final cacheKey = _getCacheKey(key);

      final item = CacheItem<T>(
        data: data,
        timestamp: DateTime.now(),
        ttl: ttl,
        key: key,
      );

      String jsonString;
      if (toJson != null) {
        jsonString = jsonEncode({
          'data': toJson(data),
          'timestamp': item.timestamp.toIso8601String(),
          'ttl': item.ttl.inMilliseconds,
          'key': item.key,
        });
      } else {
        jsonString = jsonEncode(item.toJson());
      }

      await prefs.setString(cacheKey, jsonString);
      _sets++;
      _saveStats();

      if (kDebugMode) {
        print('Cache SET: $key (TTL: ${ttl.inSeconds}s)');
      }
    } catch (e) {
      if (kDebugMode) {
        print('Cache SET error: $e');
      }
    }
  }

  /// Get an item from cache
  Future<T?> get<T>(
    String key, {
    T Function(dynamic)? fromJson,
  }) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final cacheKey = _getCacheKey(key);

      final jsonString = prefs.getString(cacheKey);
      if (jsonString == null) {
        _misses++;
        _saveStats();
        if (kDebugMode) {
          print('Cache MISS: $key');
        }
        return null;
      }

      final json = jsonDecode(jsonString);
      final item = CacheItem<T>.fromJson(json, fromJson ?? (data) => data as T);

      if (item.isExpired) {
        await prefs.remove(cacheKey);
        _misses++;
        _saveStats();
        if (kDebugMode) {
          print('Cache EXPIRED: $key');
        }
        return null;
      }

      _hits++;
      _saveStats();
      if (kDebugMode) {
        print('Cache HIT: $key');
      }
      return item.data;
    } catch (e) {
      if (kDebugMode) {
        print('Cache GET error: $e');
      }
      _misses++;
      _saveStats();
      return null;
    }
  }

  /// Check if a key exists and is not expired
  Future<bool> has(String key) async {
    final data = await get(key);
    return data != null;
  }

  /// Remove an item from cache
  Future<bool> delete(String key) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final cacheKey = _getCacheKey(key);
      final removed = await prefs.remove(cacheKey);

      if (kDebugMode) {
        print('Cache DELETE: $key');
      }
      return removed;
    } catch (e) {
      if (kDebugMode) {
        print('Cache DELETE error: $e');
      }
      return false;
    }
  }

  /// Clear all cache
  Future<void> clear() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final keys = prefs.getKeys();

      for (final key in keys) {
        if (key.startsWith(_cachePrefix)) {
          await prefs.remove(key);
        }
      }

      if (kDebugMode) {
        print('Cache CLEAR: All cache cleared');
      }
    } catch (e) {
      if (kDebugMode) {
        print('Cache CLEAR error: $e');
      }
    }
  }

  /// Get cache statistics
  Map<String, dynamic> getStats() {
    final total = _hits + _misses;
    final hitRate = total > 0 ? (_hits / total * 100).round() : 0;

    return {
      'hits': _hits,
      'misses': _misses,
      'sets': _sets,
      'total': total,
      'hitRate': hitRate,
    };
  }

  /// Load cache statistics
  Future<void> _loadStats() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final statsJson = prefs.getString(_statsKey);

      if (statsJson != null) {
        final stats = jsonDecode(statsJson);
        _hits = stats['hits'] ?? 0;
        _misses = stats['misses'] ?? 0;
        _sets = stats['sets'] ?? 0;
      }
    } catch (e) {
      if (kDebugMode) {
        print('Cache stats load error: $e');
      }
    }
  }

  /// Save cache statistics
  Future<void> _saveStats() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final stats = getStats();
      await prefs.setString(_statsKey, jsonEncode(stats));
    } catch (e) {
      if (kDebugMode) {
        print('Cache stats save error: $e');
      }
    }
  }

  /// Get cache key with prefix
  String _getCacheKey(String key) {
    return '$_cachePrefix$key';
  }

  /// Initialize cache service
  Future<void> initialize() async {
    await _loadStats();
    if (kDebugMode) {
      print('CacheService initialized');
    }
  }

  /// Preload important data
  Future<void> preloadData() async {
    if (kDebugMode) {
      print('Preloading important data...');
    }

    // This will be implemented by specific services
    // that need to preload data
  }

  /// Invalidate cache by pattern
  Future<void> invalidatePattern(String pattern) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final keys = prefs.getKeys();

      int invalidated = 0;
      for (final key in keys) {
        if (key.startsWith(_cachePrefix) && key.contains(pattern)) {
          await prefs.remove(key);
          invalidated++;
        }
      }

      if (kDebugMode) {
        print(
            'Cache INVALIDATE: $invalidated items matching pattern "$pattern"');
      }
    } catch (e) {
      if (kDebugMode) {
        print('Cache INVALIDATE error: $e');
      }
    }
  }
}

// Global cache service instance
final cacheService = CacheService();

// Cache key generators
class CacheKeys {
  // Auth related
  static String userProfile(String userId) => 'user:profile:$userId';
  static const String authToken = 'auth:token';
  static const String userData = 'auth:user';

  // Company related
  static const String companyFeatures = 'company:features';
  static String companyDetails(String companyId) =>
      'company:details:$companyId';
  static const String companiesList = 'companies:list';

  // Attendance related
  static const String todayAttendance = 'attendance:today';
  static const String attendanceSummary = 'attendance:summary';
  static String attendanceRecords(String date) => 'attendance:records:$date';

  // Leave related
  static const String leaveInfo = 'leave:info';
  static const String leaveTypes = 'leave:types';
  static String leaveRecords(String year) => 'leave:records:$year';

  // Payroll related
  static const String payrollInfo = 'payroll:info';
  static String payrollRecords(String month) => 'payroll:records:$month';

  // Settings related
  static const String appSettings = 'settings:app';
  static const String companySettings = 'settings:company';

  // Notifications
  static const String notifications = 'notifications:list';

  // Analytics
  static String analytics(String period) => 'analytics:$period';
}

File: admin-portal/src/services/cacheService.ts
// Cache Service for Admin Portal
// Provides intelligent caching to reduce API calls and improve performance

interface CacheItem<T> {
  data: T;
  timestamp: number;
  ttl: number; // Time to live in milliseconds
  key: string;
}

interface CacheConfig {
  defaultTTL: number; // Default time to live in milliseconds
  maxSize: number; // Maximum number of items in cache
  cleanupInterval: number; // How often to clean expired items
}

class CacheService {
  private cache = new Map<string, CacheItem<any>>();
  private config: CacheConfig;
  private cleanupTimer: NodeJS.Timeout | null = null;

  constructor(config: Partial<CacheConfig> = {}) {
    this.config = {
      defaultTTL: 5 * 60 * 1000, // 5 minutes default
      maxSize: 100, // Maximum 100 items
      cleanupInterval: 60 * 1000, // Cleanup every minute
      ...config
    };

    this.startCleanupTimer();
  }

  /**
   * Set an item in cache
   */
  set<T>(key: string, data: T, ttl?: number): void {
    const item: CacheItem<T> = {
      data,
      timestamp: Date.now(),
      ttl: ttl || this.config.defaultTTL,
      key
    };

    // Remove oldest items if cache is full
    if (this.cache.size >= this.config.maxSize) {
      this.evictOldest();
    }

    this.cache.set(key, item);
  }

  /**
   * Get an item from cache
   */
  get<T>(key: string): T | null {
    const item = this.cache.get(key);
    
    if (!item) {
      return null;
    }

    // Check if item has expired
    if (Date.now() - item.timestamp > item.ttl) {
      this.cache.delete(key);
      return null;
    }

    return item.data;
  }

  /**
   * Check if a key exists and is not expired
   */
  has(key: string): boolean {
    return this.get(key) !== null;
  }

  /**
   * Remove an item from cache
   */
  delete(key: string): boolean {
    return this.cache.delete(key);
  }

  /**
   * Clear all cache
   */
  clear(): void {
    this.cache.clear();
  }

  /**
   * Get cache statistics
   */
  getStats() {
    const now = Date.now();
    let expiredCount = 0;
    let validCount = 0;

    this.cache.forEach(item => {
      if (now - item.timestamp > item.ttl) {
        expiredCount++;
      } else {
        validCount++;
      }
    });

    return {
      total: this.cache.size,
      valid: validCount,
      expired: expiredCount,
      maxSize: this.config.maxSize
    };
  }

  /**
   * Evict oldest items when cache is full
   */
  private evictOldest(): void {
    let oldestKey: string | null = null;
    let oldestTime = Date.now();

    this.cache.forEach((item, key) => {
      if (item.timestamp < oldestTime) {
        oldestTime = item.timestamp;
        oldestKey = key;
      }
    });

    if (oldestKey) {
      this.cache.delete(oldestKey);
    }
  }

  /**
   * Clean up expired items
   */
  private cleanup(): void {
    const now = Date.now();
    const expiredKeys: string[] = [];

    this.cache.forEach((item, key) => {
      if (now - item.timestamp > item.ttl) {
        expiredKeys.push(key);
      }
    });

    expiredKeys.forEach(key => this.cache.delete(key));

    if (expiredKeys.length > 0) {
      console.log(`CacheService: Cleaned up ${expiredKeys.length} expired items`);
    }
  }

  /**
   * Start cleanup timer
   */
  private startCleanupTimer(): void {
    this.cleanupTimer = setInterval(() => {
      this.cleanup();
    }, this.config.cleanupInterval);
  }

  /**
   * Stop cleanup timer
   */
  destroy(): void {
    if (this.cleanupTimer) {
      clearInterval(this.cleanupTimer);
      this.cleanupTimer = null;
    }
    this.clear();
  }
}

// Cache configurations for different data types
export const CACHE_CONFIGS = {
  // Short-lived cache for frequently changing data
  SHORT: {
    defaultTTL: 30 * 1000, // 30 seconds
    maxSize: 50
  },
  
  // Medium cache for moderately changing data
  MEDIUM: {
    defaultTTL: 5 * 60 * 1000, // 5 minutes
    maxSize: 100
  },
  
  // Long cache for rarely changing data
  LONG: {
    defaultTTL: 30 * 60 * 1000, // 30 minutes
    maxSize: 50
  },
  
  // Very long cache for static data
  STATIC: {
    defaultTTL: 24 * 60 * 60 * 1000, // 24 hours
    maxSize: 20
  }
};

// Create cache instances
export const shortCache = new CacheService(CACHE_CONFIGS.SHORT);
export const mediumCache = new CacheService(CACHE_CONFIGS.MEDIUM);
export const longCache = new CacheService(CACHE_CONFIGS.LONG);
export const staticCache = new CacheService(CACHE_CONFIGS.STATIC);

// Cache key generators
export const CACHE_KEYS = {
  // Auth related
  USER_PROFILE: (userId: string) => `user:profile:${userId}`,
  AUTH_TOKEN: 'auth:token',
  
  // Company related
  COMPANIES_LIST: 'companies:list',
  COMPANY_DETAILS: (companyId: string) => `company:details:${companyId}`,
  COMPANY_FEATURES: (companyId: string) => `company:features:${companyId}`,
  
  // Subscription plans
  SUBSCRIPTION_PLANS: 'subscription:plans',
  PLAN_DETAILS: (planId: string) => `plan:details:${planId}`,
  
  // Analytics and stats
  DASHBOARD_STATS: 'dashboard:stats',
  ANALYTICS_DATA: (timeRange: string) => `analytics:${timeRange}`,
  
  // User management
  USERS_LIST: 'users:list',
  USER_DETAILS: (userId: string) => `user:details:${userId}`,
  
  // Settings
  SYSTEM_SETTINGS: 'system:settings',
  ADMIN_SETTINGS: (companyId: string) => `admin:settings:${companyId}`,
  
  // Notifications
  NOTIFICATIONS_LIST: 'notifications:list',
  
  // Monitoring
  HEALTH_STATUS: 'monitoring:health',
  ERROR_STATS: 'monitoring:errors',
  PERFORMANCE_STATS: 'monitoring:performance'
};

export default CacheService; 